aten/src/ATen/native/mkldnn/RNN.cpp

```diff
@@ -0,0 +1,583 @@
+    return at::cat({b1[1] + b2[1], b1[0] + b2[0], b1[2], b2[2]}, /*output_channels*/0);
+  }
+  return bias_ih + bias_hh;
+};
+
+// Create mkldnn memory view from ATen tensor
+static inline ideep::tensor get_mkldnn_tensor(
+    const Tensor& tensor, const ideep::tensor::desc& desc) {
+  TORCH_CHECK(
+      tensor.device().is_cpu(),
+      "get_mkldnn_tensor expects CPU tensor input");
+  TORCH_CHECK(
+      tensor.layout() == at::Layout::Strided,
+      "get_mkldnn_tensor expects dense tensor input");
+  TORCH_CHECK(
+      tensor.scalar_type() == at::ScalarType::Float ||
+          tensor.scalar_type() == at::ScalarType::BFloat16 ||
+          tensor.scalar_type() == at::ScalarType::Half,
+      "get_mkldnn_tensor expects float or bfloat16 tensor input");
+  return {desc, tensor.data_ptr()};
+}
+
+std::tuple<Tensor, Tensor, Tensor, Tensor> mkldnn_rnn_layer(const Tensor& input,
+    const Tensor& w0,
+    const Tensor& w1,
+    const Tensor& w2,
+    const Tensor& w3,
+    const Tensor& hx_,
+    const Tensor& cx_,
+    bool reverse,
+    IntArrayRef batch_sizes,
+    int64_t mode,
+    int64_t hidden_size,
+    int64_t num_layers,
+    bool has_biases,
+    bool bidirectional,
+    bool batch_first,
+    bool train) {
+  RNNParams rnn(
+      input,
+      batch_sizes,
+      mode,
+      hidden_size,
+      num_layers,
+      bidirectional,
+      batch_first,
+      train);
+
+  auto output_size = _output_size</*is_single_direction*/ true>(rnn);
+  auto output = at::empty(output_size, input.options());
+
+  auto hy_ = at::empty(hx_.sizes(), hx_.options());
+  auto cy_ = at::empty(cx_.sizes(), cx_.options());
+
+  auto weight_ih = _shuffle_weight(w0, rnn.mode);
+  auto weight_hh = _shuffle_weight(w1, rnn.mode);
+
+  auto bias = has_biases
+      ? _shuffle_bias(w2, w3, rnn.mode)
+      : at::zeros({rnn.num_bias_gates * rnn.hidden_size}, weight_ih.options());
+
+  // per layer input size
+  int64_t input_size = input.size(2);
+  auto x = get_mkldnn_tensor(
+      input,
+      rnn.src_layer_desc(input_size, get_mkldnn_dtype(input.scalar_type())));
+  auto hx = get_mkldnn_tensor(
+      hx_, rnn.src_iter_desc(get_mkldnn_dtype(hx_.scalar_type())));
+  auto cx = get_mkldnn_tensor(
+      cx_, rnn.src_iter_c_desc(get_mkldnn_dtype(cx_.scalar_type())));
+  auto b = get_mkldnn_tensor(
+      bias, rnn.bias_desc(get_mkldnn_dtype(bias.scalar_type())));
+  auto y = get_mkldnn_tensor(
+      output, rnn.dst_layer_desc(get_mkldnn_dtype(output.scalar_type())));
+  auto hy = get_mkldnn_tensor(
+      hy_, rnn.dst_iter_desc(get_mkldnn_dtype(hy_.scalar_type())));
+  auto cy = get_mkldnn_tensor(
+      cy_, rnn.dst_iter_c_desc(get_mkldnn_dtype(cy_.scalar_type())));
+  auto w1_ = get_mkldnn_tensor(weight_ih, rnn.weights_layer_desc(input_size, get_mkldnn_dtype(weight_ih.scalar_type())));
+  auto w2_ = get_mkldnn_tensor(weight_hh, rnn.weights_iter_desc(get_mkldnn_dtype(weight_hh.scalar_type())));
```

Comment:
Nit: please add `get_mkldnn_dtype` convenience function that takes `const Tensor&` as an argument.
Perhaps it'll be nicer to add it as static inline right after https://github.com/pytorch/pytorch/blob/e1bcbbf18cae2cf228111b9da784f1052070c724/aten/src/ATen/native/mkldnn/MKLDNNCommon.h#L12
rather than use lambda, but it's up to you
```suggestion
  auto get_mkldnn_dtype = [](const Tensor& t) { return get_mkldnn_dtype(t.scalar_type()); };
  auto x = get_mkldnn_tensor(
      input,
      rnn.src_layer_desc(input_size, get_mkldnn_dtype(input)));
  auto hx = get_mkldnn_tensor(
      hx_, rnn.src_iter_desc(get_mkldnn_dtype(hx_));
  auto cx = get_mkldnn_tensor(
      cx_, rnn.src_iter_c_desc(get_mkldnn_dtype(cx_)));
  auto b = get_mkldnn_tensor(
      bias, rnn.bias_desc(get_mkldnn_dtype(bias)));
  auto y = get_mkldnn_tensor(
      output, rnn.dst_layer_desc(get_mkldnn_dtype(output)));
  auto hy = get_mkldnn_tensor(
      hy_, rnn.dst_iter_desc(get_mkldnn_dtype(hy_)));
  auto cy = get_mkldnn_tensor(
      cy_, rnn.dst_iter_c_desc(get_mkldnn_dtype(cy_)));
  auto w1_ = get_mkldnn_tensor(weight_ih, rnn.weights_layer_desc(input_size, get_mkldnn_dtype(weight_ih)));
  auto w2_ = get_mkldnn_tensor(weight_hh, rnn.weights_iter_desc(get_mkldnn_dtype(weight_hh)));
```